### PR TITLE
Chore: Use exp/slices and exp/maps to simplify some code

### PR DESCRIPTION
**Pull request by jannotti**
The intent is to only make changes where the result is clearer, and to catch enough cases that the next time someone does something similar, they will have seen the new, clearer way to write the thing they intend.

### CODE REVIEW

File: agreement/autopsy.go
Comment (jannotti) [inline agreement/autopsy.go:106]: I'm not even sure this needs the clone.  When called as varargs, the slice is surely unshared.

File: ledger/apply/application_test.go
Comment (jannotti) [inline ledger/apply/application_test.go:576]: This appeared to be testing an artifact of the mock, not of a real ledger. The artifact is gone now, because `maps.Clone` maintains `nil`.

File: data/basics/teal.go
Comment (jannotti) [inline data/basics/teal.go:81]: Comment goes into a lot of detail about the `nil` checking choice, but that is exactly the same choiie that maps.Equal made.  Unit tests already exist to confirm.

File: data/transactions/transaction.go
Comment (jannotti) [reply data/transactions/transaction.go:276]: You're right.

File: crypto/merkletrie/committer_test.go
Comment (jannotti) [reply crypto/merkletrie/committer_test.go:37]: Good question.  I think they reduce to the exact same code. Part of me says we should use the specific call, because it seems like that's the point of `bytes.Clone` existing. But another part thinks the existence of `bytes.Clone` is just a historic artifact from before generics, so why write different things for different types?

What do you (or anyone else) think?

File: ledger/acctupdates.go
Comment (jannotti) [reply ledger/acctupdates.go:617]: Indeed, thank you.

But, while trying to track down why this wasn't tested, I'm getting the impression this is dead code.  There are tests that use it, but nothing else (I think).  Still looking, but I'll either remove it entirely or fix.

File: ledger/apply/mockBalances_test.go
Comment (jannotti) [reply ledger/apply/mockBalances_test.go:169]: agreed.

Comment (jannotti) [reply ledger/apply/mockBalances_test.go:169]: How about this instead? (it'll ensure that `new` is made with a size appropriate to the new bigger length.)
```
func mapWith[M ~map[K]V, K comparable, V any](m M, k K, v V) M {
	new := make(M, len(m)+1)
	maps.Copy(new, m)
	new[k] = v
	return new
}
```

File: ledger/ledgercore/statedelta.go
Comment (jannotti) [reply ledger/ledgercore/statedelta.go:557]: No, it probably doesn't. I've revert.

File: ledger/apply/mockBalances_test.go
Comment (jasonpaulos) [reply ledger/apply/mockBalances_test.go]: That looks good (though I don't love shadowing the `new` keyword)

Comment (jannotti) [reply ledger/apply/mockBalances_test.go]: ok, the linter may up yelling at me anyway, so I'll change to `newMap`

File: crypto/merkletrie/committer_test.go
Comment (jasonpaulos) [reply crypto/merkletrie/committer_test.go:37]: I can't think of a reason to have a preference for one over the other, but I wanted to have the discussion in case someone else knows of one.

FWIW `bytes.Clone` was introduced in 1.20, so definitely not pre-generics.

Comment (jannotti) [reply crypto/merkletrie/committer_test.go:37]: I just read a bit about it. Nothing conclusive, but I get the impression it only went in because strings.Clone exists, and they have some sort of pathological desire to make the API surface of those identical (even going through the trouble of adding strings.Compare and explicitly making it slow and recommending against use!)

So I'm going to say I lean toward slices.Clone everywhere. But I don't feel strongly.

File: ledger/acctupdates.go
Comment (jannotti) [reply ledger/acctupdates.go:617]: All gone now.

File: ledger/apply/application_test.go
Comment (ahangsu) [inline ledger/apply/application_test.go:48]: both `cloneAppParams` and `cloneAppLocalStates` seem to be used 2 times in this file, guess we can go further by taking these 2 down and replace both usages with `maps.Clone`?

**Review by ahangsu — APPROVED** [review]
seems fine to me, based on the assumption that `slice.Clone` has similar perf to `copy`

File: ledger/apply/application_test.go
Comment (jannotti) [reply ledger/apply/application_test.go:48]: I was tempted to do that - since the function is now one line, it seems kind of pointless.

Do any other reviewers like or dislike such a change?

Comment (jasonpaulos) [reply ledger/apply/application_test.go:48]: I'm in favor of removing the function

Comment (jannotti) [reply ledger/apply/application_test.go:48]: Two votes in favor. Sold!

**Comment by jasonpaulos** [comment]
Happy to approve once the merge conflict is fixed

**Review by ahangsu — APPROVED** [review]

File: ledger/acctupdates.go
Comment (algorandskiy) [inline ledger/acctupdates.go:649]: forgot to remove `func (qs *accountsDbQueries) ListCreatables` in sql.go and `listCreatablesStmt` there as well

Comment (jannotti) [reply ledger/acctupdates.go:649]: Thanks.
https://github.com/algorand/go-algorand/pull/5495

**Pull request merged**
